crates/simulations/src/cozy/agents/arbitrageur.rs

```diff
@@ -0,0 +1,232 @@
+    set_risk_model::SetRiskModel,
+    types::ReactionTime,
+    world::{CozyUpdate, CozyWorld},
+};
+
+#[derive(Debug, Clone)]
+pub struct ArbitrageurPreferences {
+    risk_model: SetRiskModel,
+    reaction_time: ReactionTime,
+}
+
+impl ArbitrageurPreferences {
+    pub fn new(risk_model: SetRiskModel, reaction_time: ReactionTime) -> Self {
+        Self {
+            risk_model,
+            reaction_time,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Arbitrageur {
+    _name: Cow<'static, str>,
+    address: Address,
+    protocol: Arc<ProtocolContracts>,
+    set: Arc<SetContracts>,
+    preferences: ArbitrageurPreferences,
+    rng: StdRng,
+}
+
+impl Arbitrageur {
+    pub fn new(
+        _name: Cow<'static, str>,
+        address: Address,
+        protocol: Arc<ProtocolContracts>,
+        set: Arc<SetContracts>,
+        preferences: ArbitrageurPreferences,
+        rng: StdRng,
+    ) -> Self {
+        Self {
+            _name,
+            address,
+            protocol,
+            set,
+            preferences,
+            rng,
+        }
+    }
+}
+
+impl Agent<CozyUpdate, CozyWorld> for Arbitrageur {
+    fn address(&self) -> Address {
+        self.address
+    }
+
+    fn activation_step(&mut self, state: &mut State<CozyUpdate, CozyWorld>) {
+        let router_approve_tx = self
+            .set
+            .base_token
+            .approve(self.protocol.cozy_router.address(), U256::MAX);
+        let _ = state.execute_evm_tx_and_decode(self.address, router_approve_tx);
+
+        for (_, ptoken) in self.set.ptokens.iter() {
+            let ptoken_approve_tx = ptoken.approve(self.protocol.cozy_router.address(), U256::MAX);
+            let _ = state.execute_evm_tx_and_decode(self.address, ptoken_approve_tx);
+        }
+    }
+
+    fn step(
+        &mut self,
+        state: &State<CozyUpdate, CozyWorld>,
+        channel: &AgentChannelSender<CozyUpdate>,
+    ) {
+        if !self
+            .preferences
+            .reaction_time
+            .time_to_react(state.timestamp(), &mut self.rng)
+        {
+            return;
+        }
+
+        // Get current balance.
+        let balance = state
+            .call_evm_tx_and_decode(
+                self.address,
+                self.set.base_token.balance_of(self.address.into()),
+            )
+            .expect("Error getting balance.");
+
+        // Choose market to arb.
+        let target_market = self
+            .rng
+            .gen_range(0..self.preferences.risk_model.market_weights.len());
+
+        // Compute current value of protection.
+        let (current_ptokens, current_protection_value) =
+            self.get_protection_owned(state, target_market);
+        let fair_price_percentage = self.compute_fair_cost_percentage(target_market);
+        if current_protection_value > U256::zero() {
+            // Sell if you can arbitrage.
+            let refund_value_tx = self.protocol.cozy_router.sell(
```

Comment:
could technically use previewSale here, not sure if there's benefit though

Reply:
actually, if we also wanted to factor in gas fees, it'd make more sense to use sell

Reply:
Makes sense -- `sale` might be better when we add gas fees to the picture.

One advantage of `previewSale` is that I think it does not revert (?). So, the agent could directly see the refund and decide to execute if the refund was non-zero, instead of having to check if the result of the call was a success or not. But, they would have create a new `sell` transaction, then. So, this is a little cleaner code-wise.